### PR TITLE
DOCSP-16533 extracted API links

### PR DESCRIPTION
**Pull request by kyuan-mongodb**
## Pull Request Info

Extracted API Links in situations it made sense to do so. 

Additionally:
- all the API links are :java-doc:
- API Documentation and Server Manual Entry are all capitalized 

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-16533

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-16533-ExtractAPILinks/

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

### If your page documents a concept, does it meet the following criteria?

- [ ] Target the [Jasmin persona](https://drive.google.com/file/d/14FbBOLCVxwSP6M9BK4Nz1Ir9tzxT8_02/view)
- [ ] Target the [Lucas persona](https://drive.google.com/file/d/1J2vqJxo7ldv7OP_obA9Q-avf0o_ju4Lk/view)

### CODE REVIEW

**Review by terakilobyte — APPROVED** [review]
LGTM for now. I'm afraid we've foot-gunned ourselves here though. The reason it said `sync` in the role is because there's a reactive version that will eventually have links too, and sync helps to separate them out :/

File: source/whats-new.txt [hunk 1]
```diff
@@ -25,18 +25,16 @@ New features of the 4.2 Java driver release include:
...
-- Added a `Filters.empty() <https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/model/Filters.html#empty()>`_ method
+- Added :java-docs:`MongoClients <apidocs/mongodb-driver-sync/com/mongodb/client/MongoClients.html>` instances with ``MongoClientSettings`` or ``ConnectionString`` as the configuration
+-  The ``explain()`` method can now be used on :java-docs:`find <apidocs/mongodb-driver-sync/com/mongodb/client/FindIterable.html#explain()>` and :java-docs:`aggregate <apidocs/mongodb-driver-sync/com/mongodb/client/AggregateIterable.html#explain()>` commands
+- Added a :java-docs:`JsonObject <apidocs/bson/org/bson/json/JsonObject.html>` class to make encoding from and decoding to JSON more efficient by avoiding an intermediate Map representation
+- Added a :java-docs:`BsonRepresentation <apidocs/bson/org/bson/codecs/pojo/annotations/BsonRepresentation.html>` annotation that allows ``ObjectId`` BSON values to be represented as a ``String`` in `POJO <https://en.wikipedia.org/wiki/Plain_old_Java_object>`_ classes
+- Added a :java-docs:`Filters.empty() <apidocs/mongodb-driver-core/com/mongodb/client/model/Filters.html#empty()>` method
```
Comment (biniona-mongodb) [inline source/whats-new.txt:32]: Issue: I don't think these links should be updated. These correspond to specific version of the driver, so I think the driver version should be hardcoded. This comment applies to most links on the whats new page.

File: source/fundamentals/data-formats/document-data-format-extended-json.txt [hunk 2]
```diff
@@ -33,7 +33,7 @@ Extended JSON Formats
 ---------------------
 
 MongoDB Extended JSON features different string formats to represent BSON data.
-Each of the different formats conform to the `JSON RFC <https://tools.ietf.org/html/rfc7159>`__
```
Comment (biniona-mongodb) [inline source/fundamentals/data-formats/document-data-format-extended-json.txt:36]: Issue: I do not believe this link should not be extracted

File: source/fundamentals/data-formats/document-data-format-extended-json.txt [hunk 3]
```diff
@@ -53,35 +53,31 @@ See the table below to see a description of each format:
...
+resources:
 
-For more detailed information on these formats, see the
-`Extended JSON specification <https://github.com/mongodb/specifications/blob/master/source/extended-json.rst>`__.
+- `JSON RFC <https://tools.ietf.org/html/rfc7159>`__ Official Documentation
+- :manual:`MongoDB Extended JSON </reference/mongodb-extended-json/>` Server Manual Entry
+- :manual:`Data Types in the mongo shell </core/shell-types/>` Server Manual Entry
+- :java-docs:`BsonBinary <apidocs/bson/org/bson/BsonBinary.html>` API Documentation
+- `Extended JSON specification <https://github.com/mongodb/specifications/blob/master/source/extended-json.rst>`__ GitHub Documentation 
+- :spec:`Special Rules for Parsing $uuid Fields </extended-json.rst#special-rules-for-parsing-uuid-fields>` GitHub Documentation
```
Comment (biniona-mongodb) [inline source/fundamentals/data-formats/document-data-format-extended-json.txt:80]: Issue: I do not believe these links should be extracted.

**Review by biniona-mongodb — CHANGES_REQUESTED** [review]
I think we should discuss moving non-API links out of prose.

File: source/fundamentals/crud/write-operations/delete.txt [hunk 4]
```diff
@@ -144,9 +140,14 @@ collection:
...
+see the following resources:
+
+- :java-docs:`deleteOne() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#deleteOne(org.bson.conversions.Bson)>` API Documentation 
+- :java-docs:`deleteMany() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#deleteMany(org.bson.conversions.Bson)>` API Documentation
+- :java-docs:`findOneAndDelete() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#findOneAndDelete(org.bson.conversions.Bson)>` API Documentation
+- :java-docs:`DeleteOptions <apidocs/mongodb-driver-core/com/mongodb/client/model/DeleteOptions.html>` API Documentation
+- :java-docs:`FindOneAndDeleteOptions <apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndDeleteOptions.html>` API Documentation
+- :manual:`db.collection.deleteOne() </reference/method/db.collection.deleteMany/>` Server Manual Entry
+- :manual:`db.collection.deleteMany() </reference/method/db.collection.deleteOne/>` Server Manual Entry
+- :manual:`db.collection.findOneAndDelete() </reference/method/db.collection.findOneAndDelete/>` Server Manual Entry
```
Comment (biniona-mongodb) [inline source/fundamentals/crud/write-operations/delete.txt:153]: Issue: I do not think we discussed moving manual links out of prose as a team (only API links). I believe this is a largely opinionated change that should be discussed as a group before being implemented.

**Review by biniona-mongodb — CHANGES_REQUESTED** [review]
Sorry forgot to save this comment. I think we should nail down what links are being moved out. Given how much this PR changes the entire docs, I think this is an important conversation.

File: source/fundamentals/data-formats/document-data-format-extended-json.txt (hunk 2 above)
Comment (terakilobyte) [reply source/fundamentals/data-formats/document-data-format-extended-json.txt:36]: I disagree upon review. This is a distractor link that would result in the user navigating away to a page that does not present an alternative to the information on this page.

File: source/fundamentals/builders/filters.txt [hunk 5]
```diff
@@ -49,15 +49,14 @@ types of operators:
 - :ref:`Bitwise <bitwise>`
 - :ref:`Geospatial <geospatial>`
 
-The :java-core-api:`Filters <com/mongodb/client/model/Filters.html>`
-class provides static factory methods for all the MongoDB query
+The Filters class provides static factory methods for all the MongoDB query
```
Comment (biniona-mongodb) [inline source/fundamentals/builders/filters.txt:52]:
```suggestion
The ``Filters`` class provides static factory methods for all the MongoDB query
```

Nit: I think this should receive monospace formatting

File: source/fundamentals/builders/indexes.txt [hunk 6]
```diff
@@ -31,16 +31,15 @@ the value of a specific field or set of fields, ordered by the value of
 the field. See our guide on :doc:`Indexes </fundamentals/indexes>` for
 examples of queries covered by indexes. 
 
-The :java-core-api:`Indexes <com/mongodb/client/model/Indexes.html>`
-class provides static factory methods for all the MongoDB index types.
+The Indexes class provides static factory methods for all the MongoDB index types.
```
Comment (biniona-mongodb) [inline source/fundamentals/builders/indexes.txt:34]:
```suggestion
The ``Indexes`` class provides static factory methods for all the MongoDB index types.
```

Nit: I think this receives monospace

File: source/fundamentals/builders/sort.txt [hunk 7]
```diff
@@ -169,5 +168,6 @@ method. For a detailed example showing how to specify sort criteria using
 the ``Sorts.metaTextScore()`` method, see 
 :ref:`the text search section of our sorting fundamentals guide <sorts-crud-text-search>`.
 
-For more information, see the
-:java-docs:`Sorts class API documentation </apidocs/mongodb-driver-core/com/mongodb/client/model/Sorts.html>`.
+For more information, see the :java-docs:`Sorts class
+<apidocs/mongodb-driver-core/com/mongodb/client/model/Sorts.html>`
+API Documentation .
```
Comment (biniona-mongodb) [inline source/fundamentals/builders/sort.txt:173]:
```suggestion
API Documentation.
```

File: source/fundamentals/crud/write-operations/delete.txt (hunk 4 above)
Comment (biniona-mongodb) [reply source/fundamentals/crud/write-operations/delete.txt:153]: Apologies this comment was hastily made as discussed in team zoom call. Please disregard.

File: source/fundamentals/data-formats/document-data-format-extended-json.txt [hunk 8]
```diff
@@ -53,35 +53,31 @@ See the table below to see a description of each format:
...
 .. note::
 
    The ``$uuid`` Extended JSON type is parsed from a String to a
-   :java-docs:`BsonBinary </apidocs/bson/org/bson/BsonBinary.html>`
-   object of binary subtype 4. For more information about ``$uuid`` field
-   parsing, see the
-   :spec:`special rules for parsing $uuid fields </extended-json.rst#special-rules-for-parsing-uuid-fields>`
-   section in the extended JSON specification.
-   
+   ``BsonBinary`` object of binary subtype 4.
```
Comment (biniona-mongodb) [inline source/fundamentals/data-formats/document-data-format-extended-json.txt:70]:
```suggestion
   ``BsonBinary`` object of binary subtype 4. 
or more information about ``$uuid`` field   parsing, see the   :spec:`special rules for parsing $uuid fields &lt;/extended-json.rst#special-rules-for-parsing-uuid-fields&gt;`
```

Suggestion: I do not believe this :spec: link should be extracted from this note. I believe it only makes sense in the context of this note, and I believe readers not concerned with UUID parsing will be confused by seeing it in the end of section external resources.

File: source/fundamentals/data-formats/documents.txt [hunk 9]
```diff
@@ -474,8 +483,8 @@ data from the collection using the following code:
 
    The code sample above uses helper methods that check the returned type
    and throw an exception if it is unable to cast the field value.
-   You can call the :java-docs:`get() </apidocs/bson/org/bson/BasicBSONObject.html?is-external=true#get(java.lang.String)>`
-   method to retrieve values as type type ``Object`` and to skip type checking.
+   You can call the ``get()`` method to retrieve values as type type
```
Comment (biniona-mongodb) [inline source/fundamentals/data-formats/documents.txt:486]:
```suggestion
   You can call the ``get()`` method to retrieve values as type
```

File: source/usage-examples/bulkWrite.txt [hunk 10]
```diff
@@ -92,15 +91,17 @@ The output should look something like this:
...
-- :java-docs:`InsertOneModel <apidocs/mongodb-driver-core/com/mongodb/client/model/InsertOneModel.html>`
-- :java-docs:`UpdateOneModel <apidocs/mongodb-driver-core/com/mongodb/client/model/UpdateOneModel.html>`
-- :java-docs:`UpdateManyModel <apidocs/mongodb-driver-core/com/mongodb/client/model/UpdateManyModel.html>`
-- :java-docs:`DeleteOneModel <apidocs/mongodb-driver-core/com/mongodb/client/model/DeleteOneModel.html>`
-- :java-docs:`DeleteManyModel <apidocs/mongodb-driver-core/com/mongodb/client/model/DeleteManyModel.html>`
-- :java-docs:`ReplaceOneModel <apidocs/mongodb-driver-core/com/mongodb/client/model/ReplaceOneModel.html>`
+page, see the following resources:
+
+- :manual:`Unique Index </core/index-unique>` Server Manual Entry
+- :manual:`Schema Validation </core/schema-validation>` Server Manual Entry
```
Comment (biniona-mongodb) [inline source/usage-examples/bulkWrite.txt:97]: Comment: This is the first instance of changing manual links to an end of section I have seen in this PR. Apologies for thinking this was a wider reaching theme in this PR than it was in reality. I panicked when I thought this PR made this change throughout the whole documentation, and I didn't think clearly. I will try and not have this alarmist knee-jerk reaction in the future.

File: source/whats-new.txt (hunk 1 above)
Comment (biniona-mongodb) [inline source/whats-new.txt:32]: Issue: as these links correspond to specific versions of the driver. I believe they should be hardcoded as before.

File: source/whats-new.txt [hunk 11]
```diff
@@ -25,18 +25,16 @@ New features of the 4.2 Java driver release include:
...
 
 .. important::
 
    There are breaking changes that may affect your application. See the
-   `Upgrading Guide
-   <https://mongodb.github.io/mongo-java-driver/4.2/upgrading/>`_ for
-   more information.
+   :java-docs:`Upgrading Guide <upgrading/>` for more information.
```
Comment (biniona-mongodb) [inline source/whats-new.txt:37]: Issue: I believe this link was correctly hardcoded before.

**Review by biniona-mongodb — APPROVED** [review]
Apologies for holding this up. Thank you for putting up with my nervousness, I will try not to have alarmist reactions going forward. A few small nits. 

The only item that I have concrete reason to change are the "What's New" links, as I believe they are tied to a specific driver version.

**Pull request merged**